group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,419 @@
+        this.targetAssignment = new HashMap<>();
+        // Without rack-aware strategy, tracking current owners of unassigned partitions is unnecessary
+        // as all sticky partitions are retained until a member meets its quota.
+        this.currentPartitionOwners = rackInfo.useRackStrategy ? new HashMap<>() : Collections.emptyMap();
+    }
+
+    /**
+     * Here's the step-by-step breakdown of the assignment process:
+     *
+     * <li> Compute the quotas of partitions for each member based on the total partitions and member count.</li>
+     * <li> Initialize unassigned partitions to all the topic partitions and
+     *      remove partitions from the list as and when they are assigned.</li>
+     * <li> For existing assignments, retain partitions based on the determined quota and member's rack compatibility.</li>
+     * <li> If a partition's rack mismatches with its owner, track it for future use.</li>
+     * <li> Identify members that haven't fulfilled their partition quota or are eligible to receive extra partitions.</li>
+     * <li> Proceed with a round-robin assignment adhering to rack awareness.
+     *      For each unassigned partition, locate the first compatible member from the potentially unfilled list.</li>
+     * <li> If no rack-compatible member is found, revert to the tracked current owner.
+     *      If that member can't accommodate the partition due to quota limits, resort to a generic round-robin assignment.</li>
+     */
+    @Override
+    protected GroupAssignment buildAssignment() throws PartitionAssignorException {
+        int totalPartitionsCount = 0;
+
+        for (Uuid topicId : subscriptionIds) {
+            int partitionCount = subscribedTopicDescriber.numPartitions(topicId);
+            if (partitionCount == -1) {
+                throw new PartitionAssignorException(
+                    "Members are subscribed to topic " + topicId + " which doesn't exist in the topic metadata."
+                );
+            } else {
+                totalPartitionsCount += partitionCount;
+            }
+        }
+
+        if (subscriptionIds.isEmpty()) {
+            LOG.debug("The subscription list is empty, returning an empty assignment");
+            return new GroupAssignment(Collections.emptyMap());
+        }
+
+        // The minimum required quota that each member needs to meet for a balanced assignment.
+        // This is the same for all members.
+        final int numberOfMembers = assignmentSpec.members().size();
+        final int minQuota = totalPartitionsCount / numberOfMembers;
+        remainingMembersToGetAnExtraPartition = totalPartitionsCount % numberOfMembers;
+
+        assignmentSpec.members().keySet().forEach(memberId ->
+            targetAssignment.put(memberId, new MemberAssignment(new HashMap<>())
+        ));
+
+        // Sorted list of all partitions.
+        unassignedPartitions = new HashSet<>(allTopicIdPartitions(subscriptionIds, subscribedTopicDescriber));
```

Comment:
In `allTopicIdPartitions`, we sort the list and then put it in a HashSet. Could we optimize this?